### PR TITLE
[5.0] [Guided Tours] Add context to core tours

### PR DESCRIPTION
**Pull request by obuisard**
Pull Request for Issues #40023 and #41273.

### Summary of Changes

Addition of context to the core tours.

### Testing Instructions

Check the list of tours on different pages of the administration. 
When going to 'Articles', the list of tours should present the article and the categories tours in priority.

Special case with categories:
Articles and Categories tours are now showing in each other's context.

Create a duplicate of the categories tour.
Change the title to 'Contact category tour'.
Change the url to `administrator/index.php?option=com_categories&view=categories&extension=com_contact`.
In component selector, select 'categories' and 'contacts'.
Publish the tour and save.

Go to the contacts tour.
Add the 'categories' to the component selector. Save.

You should now have the new category tour available to the contacts and contact categories as tours that show in priority in the list of tours. 
Check the articles and article categories page to make sure the new category tour for contacts is NOT present.
Check the feeds category page to make sure no tour has priority there.

### Actual result BEFORE applying this Pull Request

The numbers of tours showing in the dropdown can go up to X items, X being set in the tours administrator module.
Only the guided tours have context and show first in the list of tours when the user is listing tours or steps.
The tours follow the order in which they have been ordered in the Guided Tours component.

### Expected result AFTER applying this Pull Request

On any given page where the module 'Take a tour' is showing, the tours being in the context of the page are the ones showing in priority.
The tours are still following the order that is set in the Guided Tours component. The only difference is that now there is a pre-list of tours, more easily in reach when a tour exists for the context the user is in.
A tour having the parameter 'extensions' set to 'all' will always show in the top list.
Then are shown the next X tours, X being specified in the administrator module for the tours. Here X is set to 5. The contextual  tours are not part of the count.

image

If X is set to 0, only the contextual tours will show, which gives administrators the possibility to only show contextual tours in the list of tours.

image

### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: 
[Admin Module: Guided Tours](https://docs.joomla.org/Help5.x:Admin_Modules:_Guided_Tours)
[Guided Tours: New or Edit Tour](https://docs.joomla.org/Help5.x:Guided_Tours:_New_or_Edit_Tour)
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by brianteeman** [comment]
The markup for this is invalid as you have the `<hr>` seperator as a listitem so it is included in the count of list items but there is nothing to anounce.

What you are actually creating is multiple lists and should be marked up a such.

This [stackoverflow]([](https://stackoverflow.com/questions/69021634/the-correct-way-to-add-separators-in-lists)) desccribes it perfectly

**Comment by micker** [comment]
@obuisard bump me when is was ready for testing

**Comment by brianteeman** [comment]
You realise that with this change the articles category tour is now emphasised in every component that supports categories?

**Comment by obuisard** [comment]
> The markup for this is invalid as you have the `<hr>` seperator as a listitem so it is included in the count of list items but there is nothing to anounce.
> 
> What you are actually creating is multiple lists and should be marked up a such.
> 
> This [stackoverflow](https://stackoverflow.com/questions/69021634/the-correct-way-to-add-separators-in-lists) desccribes it perfectly

This is how the markup is documented at Bootstrap (https://getbootstrap.com/docs/5.2/components/dropdowns/#dividers)... So I guess I should take it as a grain of salt and I could separate the lists... Or I'll do like with the sidebar menu and have the li with class 'divider' and role 'presentation'.

**Comment by obuisard** [comment]
> You realise that with this change the articles category tour is now emphasised in every component that supports categories?

It does. it's a problem with categories. The category tour applies to articles in particular but it's the same tour for all components. If I only give the category tour a context 'com_categories', it still shows under 'Banner' categories. The only way this does not happen is to have specific code in the module that handles categories differently. I will just make the module 'smarter' so it checks the context and when the context is categories, check the extension it applies to (I have the info in the URL, which is all good).

**Comment by obuisard** [comment]
Brian @brianteeman I put the markup back as it was prior to this PR. I have just added the buttons at the top and hr between the sections. The ul/li was 'breaking' the UI but now I want to make sure we nailed the accessibility of the dropdown, especially when it comes to the hr tag.

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 1]
```diff
@@ -27,17 +27,44 @@
...
-
     $uri = new Uri($tour->url);
+    if (count(array_intersect(['*', $extension], $tour->extensions))) :
+        // Special case for the categories page, where the context is complemented with the extension the categories apply to.
+        if ($extension === 'com_categories') :
+            if ($uri->getVar('option', '') === 'com_categories') :
+                if ($uri->getVar('extension', '') === $app->getInput()->get('extension', '')) :
+                    $contextTours[] = $tour;
+                else :
+                    if ($toursCount > 0) :
```
Comment (Quy) [inline administrator/modules/mod_guidedtours/tmpl/default.php]: Change to `elseif` to reduce level indentation?

Comment (obuisard) [reply administrator/modules/mod_guidedtours/tmpl/default.php]: Good catch :-)

**Comment by brianteeman** [comment]
making it a list was an improvement

**Comment by obuisard** [comment]
OK Brian @brianteeman, I made an attempt with lists. Tired to also change the icon for contextual tours, not sure it adds much value.

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 2]
```diff
@@ -27,18 +27,39 @@
 
 $lang       = $app->getLanguage();
 $extension  = $app->getInput()->get('option');
+$contextTours  = [];
```
Comment (Quy) [inline administrator/modules/mod_guidedtours/tmpl/default.php]:
```suggestion
$contextTours = [];
```

Comment (obuisard) [reply administrator/modules/mod_guidedtours/tmpl/default.php]: Thanks for pointing this out, realized all needed to be aligned

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 3]
```diff
@@ -64,12 +85,32 @@
...
+                    </li>
+                <?php endforeach; ?>
+            </ol>
+            <hr class="dropdown-divider m-0" role="separator" />
+        <?php endif; ?>
+        <?php if (count($listTours) > 0) : ?>
+            <ol class="list-unstyled m-0">
+                <?php foreach ($listTours as $tour) : ?>
+                    <li>
+                        <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id ?>">
```
Comment (Quy) [inline administrator/modules/mod_guidedtours/tmpl/default.php]:
```suggestion
                        <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id; ?>">
                            <span class="icon-star" aria-hidden="true"></span>
                            <?php echo $tour->title; ?>
                        </button>
                    </li>
                <?php endforeach; ?>
            </ol>
            <hr class="dropdown-divider m-0" role="separator" />
        <?php endif; ?>
        <?php if (count($listTours) > 0) : ?>
            <ol class="list-unstyled m-0">
                <?php foreach ($listTours as $tour) : ?>
                    <li>
                        <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id; ?>">
```

Comment (obuisard) [reply administrator/modules/mod_guidedtours/tmpl/default.php]: Thanks!

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 4]
```diff
@@ -64,12 +85,32 @@
...
+                <?php endforeach; ?>
+            </ol>
+            <hr class="dropdown-divider m-0" role="separator" />
+        <?php endif; ?>
+        <?php if (count($listTours) > 0) : ?>
+            <ol class="list-unstyled m-0">
+                <?php foreach ($listTours as $tour) : ?>
+                    <li>
+                        <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id; ?>">
+                            <span class="icon-map-signs" aria-hidden="true"></span>
```
Comment (Quy) [inline administrator/modules/mod_guidedtours/tmpl/default.php]:
```suggestion
                            <span class="icon-star icon-fw" aria-hidden="true"></span>
                            <?php echo $tour->title; ?>
                        </button>
                    </li>
                <?php endforeach; ?>
            </ol>
            <hr class="dropdown-divider m-0" role="separator" />
        <?php endif; ?>
        <?php if (count($listTours) > 0) : ?>
            <ol class="list-unstyled m-0">
                <?php foreach ($listTours as $tour) : ?>
                    <li>
                        <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id; ?>">
                            <span class="icon-map-signs icon-fw" aria-hidden="true"></span>
```

**Comment by richard67** [comment]
@obuisard Could you rename the update SQL script to something newer than "5.0.0-2023-08-26.sql", e.g. "5.0.0-2023-08-27.sql" or "5.0.0-2023-08-28.sql"? Thanks in advance.

**Comment by obuisard** [comment]
> @obuisard Could you rename the update SQL script to something newer than "5.0.0-2023-08-26.sql", e.g. "5.0.0-2023-08-27.sql" or "5.0.0-2023-08-28.sql"? Thanks in advance.

Done :-) Thanks Richard.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on d56229062c8490b3bba29a6e60bed8031aaf4ec8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41303">issues.joomla.org/tracker/joomla-cms/41303</a>.</sub>

**Comment by richard67** [comment]
@obuisard It seems that with the last branch update to 5.0-dev you have merged the "5.0.0-2023-08-28.sql" update SQL scripts which had been meanwhile added with PR #41499 to the 5.0-dev branch with your.

This is ok as long as we don't have beta 1 released, because the "5.0.0-2023-08-28.sql" from PR #41499 also hasn't been released yet, so for updating from alpha 4 to beta 1 it is ok to have both things (that PR and yours) in the same script.

But as soon as beta 1 is out without your PR here, and this PR here would go into beta 2, you will have to use a new script with a newer schema version in its name.

For the same reason, when you now update a 5.0-dev to the patched update package or the custom update URL of this PR here, the update SQL would not run because the schema version is already "5.0.0-2023-08-28".

No need to fix that now, but if that happens again in future: The correct fix would be to revert your changes in the "5.0.0-2023-08-28.sql" scripts and then create new ones with a higher schema version in their name which contain only your stuff. The problem could have been avoided by renaming your scripts before doing the branch update.

I thought I let you know that.

**Comment by obuisard** [comment]
> @obuisard It seems that with the last branch update to 5.0-dev you have merged the "5.0.0-2023-08-28.sql" update SQL scripts which had been meanwhile added with PR #41499 to the 5.0-dev branch with your.
> 
> This is ok as long as we don't have beta 1 released, because the "5.0.0-2023-08-28.sql" from PR #41499 also hasn't been released yet, so for updating from alpha 4 to beta 1 it is ok to have both things (that PR and yours) in the same script.
> 
> But as soon as beta 1 is out without your PR here, and this PR here would go into beta 2, you will have to use a new script with a newer schema version in its name.
> 
> For the same reason, when you now update a 5.0-dev to the patched update package or the custom update URL of this PR here, the update SQL would not run because the schema version is already "5.0.0-2023-08-28".
> 
> No need to fix that now, but if that happens again in future: The correct fix would be to revert your changes in the "5.0.0-2023-08-28.sql" scripts and then create new ones with a higher schema version in their name which contain only your stuff. The problem could have been avoided by renaming your scripts before doing the branch update.

Thank Richard @richard67 for the explanation, very helpful. I am hoping to get this PR into beta 1.
As I understand:
- if this PR goes to beta 1, I can leave it as is,
- if the PR does NOT go into beta 1, I need to change the file names to a later date, without (of course) the extra code from the other PR.

**Comment by richard67** [comment]
> Thank Richard @richard67 for the explanation, very helpful. I am hoping to get this PR into beta 1. As I understand:
> 
>     * if this PR goes to beta 1, I can leave it as is,
> 
>     * if the PR does NOT go into beta 1, I need to change the file names to a later date, without (of course) the extra code from the other PR.

Correct.

**Comment by khu5h1** [comment]
I have tested this item :white_check_mark: successfully on 27fb3194bd2fab0e768c56961e1ac84d5c729517<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41303">issues.joomla.org/tracker/joomla-cms/41303</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 27fb3194bd2fab0e768c56961e1ac84d5c729517<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41303">issues.joomla.org/tracker/joomla-cms/41303</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41303">issues.joomla.org/tracker/joomla-cms/41303</a>.</sub>

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 5]
```diff
@@ -64,12 +101,32 @@
...
     </button>
     <div class="dropdown-menu dropdown-menu-end">
-        <?php foreach ($listTours as $tour) : ?>
-            <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id ?>">
-                <span class="icon-map-signs" aria-hidden="true"></span>
-                <?php echo $tour->title; ?>
-            </button>
-        <?php endforeach; ?>
+        <?php if (count($contextTours) > 0) : ?>
+            <ol class="list-unstyled m-0">
```
Comment (brianteeman) [inline administrator/modules/mod_guidedtours/tmpl/default.php]:
```suggestion
            <ul class="list-unstyled m-0">
```

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 6]
```diff
@@ -64,12 +101,32 @@
...
+                <?php foreach ($contextTours as $tour) : ?>
+                    <li>
+                        <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id; ?>">
+                            <span class="icon-star icon-fw" aria-hidden="true"></span>
+                            <?php echo $tour->title; ?>
+                        </button>
+                    </li>
+                <?php endforeach; ?>
+            </ol>
```
Comment (brianteeman) [inline administrator/modules/mod_guidedtours/tmpl/default.php]:
```suggestion
            </ul>
```

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 7]
```diff
@@ -64,12 +101,32 @@
...
+            <hr class="dropdown-divider m-0" role="separator" />
+        <?php endif; ?>
+        <?php if (count($listTours) > 0) : ?>
+            <ol class="list-unstyled m-0">
```
Comment (brianteeman) [inline administrator/modules/mod_guidedtours/tmpl/default.php]:
```suggestion
            <ul class="list-unstyled m-0">
```

File: administrator/modules/mod_guidedtours/tmpl/default.php [hunk 8]
```diff
@@ -64,12 +101,32 @@
...
+                <?php foreach ($listTours as $tour) : ?>
+                    <li>
+                        <button type="button" class="button-start-guidedtour dropdown-item" data-id="<?php echo $tour->id; ?>">
+                            <span class="icon-map-signs icon-fw" aria-hidden="true"></span>
+                            <?php echo $tour->title; ?>
+                        </button>
+                    </li>
+                <?php endforeach; ?>
+            </ol>
```
Comment (brianteeman) [inline administrator/modules/mod_guidedtours/tmpl/default.php]:
```suggestion
            </ul>
```

**Comment by brianteeman** [comment]
(sorry for the late response)

These lists should be ul not ordered lists. You only use ordered lists when the data is in sequence (reference https://accessibleweb.com/question-answer/what-is-an-unordered-list-vs-an-ordered-list/)

**Comment by obuisard** [comment]
> (sorry for the late response)
> 
> These lists should be ul not ordered lists. You only use ordered lists when the data is in sequence (reference https://accessibleweb.com/question-answer/what-is-an-unordered-list-vs-an-ordered-list/)

I have set it that way because the tours are actually ordered, following the order of the administrator's tours view. Do you still think I should change this Brian @brianteeman ?

**Comment by brianteeman** [comment]
Yes. See the explanation in the link I provided

**Comment by obuisard** [comment]
> Yes. See the explanation in the link I provided

I did, thanks Brian.

**Comment by sdwjoomla** [comment]
Thanks @obuisard

**Pull request merged**